src/v/cluster/archival/retention_calculator.cc

```diff
@@ -175,19 +176,25 @@ std::optional<retention_calculator> retention_calculator::factory(
         return std::nullopt;
     }
 
-    return retention_calculator{manifest, std::move(strats)};
+    return retention_calculator{manifest, std::move(strats), pinned_offset};
 }
 
 retention_calculator::retention_calculator(
   const cloud_storage::partition_manifest& manifest,
-  std::vector<std::unique_ptr<retention_strategy>> strategies)
+  std::vector<std::unique_ptr<retention_strategy>> strategies,
+  std::optional<kafka::offset> pinned_offset)
   : _manifest(manifest)
-  , _strategies(std::move(strategies)) {}
+  , _strategies(std::move(strategies))
+  , _pinned_offset(pinned_offset) {}
 
 std::optional<model::offset> retention_calculator::next_start_offset() {
     auto it = _manifest.first_addressable_segment();
     for (; it != _manifest.end(); ++it) {
         const auto& entry = *it;
+        if (_pinned_offset && entry.last_kafka_offset() >= *_pinned_offset) {
```

Comment:
ditto
two subsystems are tied together via parameter and the behaviour is more complicated and nuanced

Reply:
Same here, can you clarify? Not sure how this would work